src/data_structures/stack.rs

```diff
@@ -0,0 +1,158 @@
+use std::fmt::{self, Display, Formatter};
+use std::ptr;
+use std::ptr::NonNull;
+
+struct Node<T> {
+    val: T,
+    next: Option<NonNull<Node<T>>>,
+}
+
+impl<T> Node<T> {
+    fn new(t: T) -> Node<T> {
+        Node { val: t, next: None }
+    }
+}
+
+#[derive(Default)]
+pub struct Stack<T> {
+    head: Option<NonNull<Node<T>>>,
+    top: usize,
+}
+
+impl<T> Stack<T> {
+    pub fn new() -> Self {
+        Self { head: None, top: 0 }
+    }
+
+    /// Push an item to the top of a stack.
+    pub fn push(&mut self, value: T) {
+        let mut node = Box::new(Node::new(value));
```

Comment:
This will be in the heap rather than stack.

Reply:
what are you suggesting then?  because, as I know there should be a push method on stack right?

Reply:
I mean we are working with stack, but we are using the heap, or it is just the stack data stucture but stuff works in heap?

Reply:
sorry for the confusion, it is a stack data structure which stores data on the heap.